1-raw-data/mailinglists/getMailData.go

```diff
@@ -0,0 +1,190 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+//TODO
+// Create tests
+// check the most recent file stored and pull only what isn't there
+// Run this monthly at start of new month to pull all new data
+
+import (
+	"cloud.google.com/go/storage"
+	"context"
+	"flag"
+	"fmt"
+	"github.com/PuerkitoBio/goquery"
+	"google.golang.org/api/iterator"
+	"io"
+	"log"
+	"net/http"
+	"strings"
+)
+
+type GCS struct {
+	bucketName string
+	ctx        context.Context
+	client     *storage.Client
+	bucket     *storage.BucketHandle
+}
+
+func connectCTX() context.Context {
+	ctx := context.Background()
+	//ctx, cancel := context.WithTimeout(ctx, time.Second*10)
+	//defer cancel()
+	return ctx
+}
+
+func (gcs *GCS) connectGCS() error {
+	client, err := storage.NewClient(gcs.ctx)
+	if err != nil {
+		return fmt.Errorf("Failed to create client: %v", err)
+	}
+	gcs.client = client
+	return nil
+}
+
+func (gcs *GCS) createGCSBucket(projectID string) error {
+	// Setup client bucket to work from
+	gcs.bucket = gcs.client.Bucket(gcs.bucketName)
+
+	buckets := gcs.client.Buckets(gcs.ctx, projectID)
+	for {
+		attrs, err := buckets.Next()
+		// Assuming that if end Iterator then not found and need to create
+		if err == iterator.Done {
+			// Create bucket if it doesn't exist - https://cloud.google.com/storage/docs/reference/libraries
+			if err := gcs.bucket.Create(gcs.ctx, projectID, &storage.BucketAttrs{
+				Location: "US",
+			}); err != nil {
+				// TODO - add random number to append to bucket name to resolve
+				return fmt.Errorf("Failed to create bucket:", err)
+			}
+			log.Printf("Bucket %v created.\n", gcs.bucketName)
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("Issues setting up Bucket(%q).Objects(): %v. Double check project id.", attrs.Name, err)
+		}
+		if attrs.Name == gcs.bucketName {
+			//getLatestFile() // TODO set this up to check and compare what is in the bucket vs what isn't
```

Comment:
clean this up if ya dont need it

Reply:
Keeping for reference until check is written